src/backend/cpu/kernel/moments.hpp

```diff
@@ -0,0 +1,128 @@
+/*******************************************************
+ * Copyright (c) 2015, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#pragma once
+#include <af/defines.h>
+#include <Array.hpp>
+#include <utility.hpp>
+#include <math.hpp>
+
+namespace cpu
+{
+namespace kernel
+{
+
+
+template<typename T, af_moment MM>
+struct moments_op
+{
+    T operator()(T const * const in,  af::dim4 const & idims,
+                 dim_t const iElems, af::dim4 const & istrides, bool const pBatch,
+                 dim_t const idx, dim_t const idy, dim_t const idz, dim_t const idw)
+    {
+        return;
+    }
+};
+
+template<typename T>
+struct moments_op<T, M00>
+{
+    T operator()(T const * const in,  af::dim4 const & idims,
+                 dim_t const iElems, af::dim4 const & istrides, bool const pBatch,
+                 dim_t const idx, dim_t const idy, dim_t const idz, dim_t const idw)
+    {
+        dim_t mId = idy * istrides[1] + idx;
+        if(pBatch) {
+            mId += idw * istrides[3] + idz * istrides[2];
+        }
+
+        return in[mId];
+    }
+};
+
+template<typename T>
+struct moments_op<T, M01>
+{
+    T operator()(T const * const in,  af::dim4 const & idims,
+                 dim_t const iElems, af::dim4 const & istrides, bool const pBatch,
+                 dim_t const idx, dim_t const idy, dim_t const idz, dim_t const idw)
+    {
+        dim_t mId = idy * istrides[1] + idx;
+        if(pBatch) {
+            mId += idw * istrides[3] + idz * istrides[2];
+        }
+
+        return idx * in[mId];
+    }
+};
+
+template<typename T>
+struct moments_op<T, M10>
+{
+    T operator()(T const * const in,  af::dim4 const & idims,
+                 dim_t const iElems, af::dim4 const & istrides, bool const pBatch,
+                 dim_t const idx, dim_t const idy, dim_t const idz, dim_t const idw)
+    {
+        dim_t mId = idy * istrides[1] + idx;
+        if(pBatch) {
+            mId += idw * istrides[3] + idz * istrides[2];
+        }
+
+        return idy * in[mId];
+    }
+};
+
+template<typename T>
+struct moments_op<T, M11>
+{
+    T operator()(T const * const in,  af::dim4 const & idims,
+                 dim_t const iElems, af::dim4 const & istrides, bool const pBatch,
+                 dim_t const idx, dim_t const idy, dim_t const idz, dim_t const idw)
+    {
+        dim_t mId = idy * istrides[1] + idx;
+        if(pBatch) {
+            mId += idw * istrides[3] + idz * istrides[2];
+        }
+
+        return idx * idy * in[mId];
+    }
+};
```

Comment:
Don't pass in strides. Calculate idx, idy, idz, idw outside and pass those along aling side mId.